### PR TITLE
feat: Support AWS 4.x Terraform provider

### PR DESCRIPTION
**Pull request by shashidhar087**
Fixes #1739 

Terraform released new version of AWS provider 4.0.0 today one hour ago.  The version mentioned in our terraform is version: ">= 3.38". terraform init will install AWS version 4.0.0.  The version 4.0.0 is a breaking change for S3 resource created in "runner-binaries-syncer" module. Some of the arguments like "acl", "lifecycle_rule", "lifecycle_configuration" became "read_only" in new version.

Reference - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket

### CODE REVIEW

**Comment by shashidhar087** [comment]
The example workflows are failing since the terraform init is reusing the version present terraform.lock.hcl file .
How do we make it work ? I could think of two possible ways 

1. "terraform init -upgrade" 
2. Mention the AWS version >=4.0.0 in all the providers files

**Comment by npalm** [comment]
@shashidhar087 thanks for your contribution. We will check the PR in next days. I normally prefer to wait a bit with upgrading the new major version. Normally prefer to wait for a first bug fix. For the time being I have created a quick fix to limit the max provider version, see #1741

**Comment by npalm** [comment]
@shashidhar087 can you run a rebase. And can you run on the terraform examples `terraform init -upgrade` to upgrade the aws provider in the lock file. That should solve the breaking builds

**Comment by shashidhar087** [comment]
@npalm Thanks for reaching out. I have done rebase and resolved the conflicts.
Could you please review and let me know if any changes to be done ?

**Comment by npalm** [comment]
@shashidhar087 will check next week, builds are now all good

File: modules/runner-binaries-syncer/main.tf [hunk 1]
```diff
@@ -4,44 +4,55 @@ locals {
...
+
+  }
+}
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "action_dist" {
+  bucket = aws_s3_bucket.action_dist.id
+
+
+  dynamic "rule" {
+    for_each = length(keys(lookup(var.server_side_encryption_configuration, "rule", {}))) == 0 ? [] : [lookup(var.server_side_encryption_configuration, "rule", {})]
```
Comment (julada) [inline modules/runner-binaries-syncer/main.tf:41]:
```suggestion
  count = length(keys(lookup(var.server_side_encryption_configuration, "rule", {}))) == 0 ? 0 : 1

  dynamic "rule" {
    for_each = length(keys(lookup(var.server_side_encryption_configuration, "rule", {}))) == 0 ? [] : [lookup(var.server_side_encryption_configuration, "rule", {})]
```

  dynamic "rule" {
    for_each = [lookup(var.server_side_encryption_configuration, "rule", {})]

it is not allowed to define a `aws_s3_bucket_server_side_encryption_configuration` resource without any rule:
`aws_s3_bucket_server_side_encryption_configuration`

the resource must be omitted completely when no rule ist defined

**Review by julada — CHANGES_REQUESTED** [review]

File: modules/runner-binaries-syncer/main.tf [hunk 2]
```diff
@@ -4,44 +4,55 @@ locals {
...
+        lookup(rule.value, "apply_server_side_encryption_by_default", {})]
+
+        content {
+          sse_algorithm     = apply_server_side_encryption_by_default.value.sse_algorithm
+          kms_master_key_id = lookup(apply_server_side_encryption_by_default.value, "kms_master_key_id", null)
+        }
+      }
+    }
   }
 }
```
Comment (julada) [inline modules/runner-binaries-syncer/main.tf:57]:
```suggestion
resource "aws_s3_bucket_server_side_encryption_configuration" "action_dist" {
  bucket = aws_s3_bucket.action_dist.id

  count = length(keys(lookup(var.server_side_encryption_configuration, "rule", {}))) == 0 ? 0 : 1

  rule {
    bucket_key_enabled = lookup(var.server_side_encryption_configuration.rule, "bucket_key_enabled", null)

    dynamic "apply_server_side_encryption_by_default" {
      for_each = [var.server_side_encryption_configuration.rule.apply_server_side_encryption_by_default]

      content {
        sse_algorithm     = apply_server_side_encryption_by_default.value.sse_algorithm
        kms_master_key_id = lookup(apply_server_side_encryption_by_default.value, "kms_master_key_id", null)
      }
    }

  }
}
```

File: modules/runner-binaries-syncer/main.tf (hunk 1 above)
Comment (julada) [reply modules/runner-binaries-syncer/main.tf:41]: @shashidhar087

Comment (shashidhar087) [reply modules/runner-binaries-syncer/main.tf:41]: Good catch, thans @julada I will commit the suggested change

File: modules/runner-binaries-syncer/main.tf (hunk 2 above)
Comment (shashidhar087) [reply modules/runner-binaries-syncer/main.tf]: @julada updated resource as suggested. Thank You.

**Comment by npalm** [comment]
@shashidhar087 sorry for keep you wating. Will test the PR asap.

**Comment by prashanthbgoud** [comment]
@npalm did you get any chance to review this?

**Comment by npalm** [comment]
@shashidhar087 sorry no, was ill.

File: .github/workflows/terraform.yml [hunk 3]
```diff
@@ -55,7 +55,7 @@ jobs:
     steps:
       - uses: actions/checkout@v2
       - name: terraform init
-        run: terraform init -get -backend=false -input=false
+        run: terraform init -upgrade -get -backend=false -input=false
```
Comment (npalm) [inline .github/workflows/terraform.yml:58]: Can you revert the upgrade here, and update the examples by updating the version.tf files and running an init upgrade?

**Review by npalm — CHANGES_REQUESTED** [review]
@shashidhar087 First of my excuse for keep you wainting that long. But thanks for the great work. And we think it is time to upgrade provider. I have added one small command related to the examples. Can you update the examples as well.

**Review by shashidhar087 — COMMENTED** [review]
Review request

**Comment by shashidhar087** [comment]
@npalm - Terraform error - Could not retrieve the list of available versions for provider
│ hashicorp/aws: locked provider registry.terraform.io/hashicorp/aws 3.71.0
│ does not match configured version constraint >= 3.63.0, ~> 4.0; must use
│ terraform init -upgrade to allow selection of new versions

**Comment by npalm** [comment]
We have the lock file in the repo, we prefer an immutal setup. Please can you run in each example dir `terrraform init -upgrade` and commit the lock file. That should fix the CI

**Comment by shashidhar087** [comment]
> We have the lock file in the repo, we prefer an immutal setup. Please can you run in each example dir `terrraform init -upgrade` and commit the lock file. That should fix the CI

Done. Waiting for workflow approval.

**Review by npalm — APPROVED** [review]
LGTM, @ScottGuymer can you a double check due to the impact of the change

**Comment by npalm** [comment]
@shashidhar087 thanks, CI is happy. I have created an Issue #1938  (and pinned the issue). Let give the community several days to react. Also asked for a second review due to the impact. Thanks for you contribution!

**Comment by npalm** [comment]
@shashidhar087 please can you resolve the conflict? You can simply ignore the incoming change.

**Comment by shashidhar087** [comment]
> @shashidhar087 please can you resolve the conflict? You can simply ignore the incoming change.

Conflict resolved.

**Comment by spider-network** [comment]
Really cool, we are waiting for the AWS provider 4.x support.

**Comment by prashanthbgoud** [comment]
@npalm I am waiting for this feature, do we have any timelines for merging this?

**Comment by npalm** [comment]
> @npalm I am waiting for this feature, do we have any timelines for merging this?

I have planned some time tomorrow to get the PR's related version 4.0 merged. So hope to have the feature available this week.

**Review by npalm — APPROVED** [review]

**Pull request merged**